File: 백준 2178 미로 탐색/복습/Main.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {

    private static int N; //1차원
    private static int M; //2차원
    private static int[][] map;
    private static boolean[][] isVisited;
    private static int[] dx = {-1, 0, 1, 0};
    private static int[] dy = {0, -1, 0, 1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        map = new int[N + 2][M + 2];
        isVisited = new boolean[N + 2][M + 2];
        for (int i = 0; i < N; i++) {
            String line = sc.next();
            for (int j = 0; j < line.length(); j++) {
                int num = line.charAt(j) - '0';
                map[i + 1][j + 1] = num;
            }
        }
        bfs(1, 1);
        System.out.println(map[N][M]);
    }

    static void bfs(int x, int y) {
        Queue<Point> queue = new LinkedList<>();
        queue.offer(new Point(x, y));
        isVisited[x][y] = true;
        while (!queue.isEmpty()) {
            Point point = queue.poll();
            for (int i = 0; i < 4; i++) {
                int x2 = point.x + dx[i];
                int y2 = point.y + dy[i];
                if (map[x2][y2] == 1 && !isVisited[x2][y2]) {
                    isVisited[x2][y2] = true;
                    queue.offer(new Point(x2, y2));
                    map[x2][y2] = map[point.x][point.y] + 1;
                }
            }
        }
    }

    static class Point {
        int x;
        int y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }


}

File: 백준 2178 미로 탐색/복습/Main.kt
import java.util.*


private var N: Int = 0 //1차원
private var M: Int = 0 //2차원
private var map: Array<IntArray>? = null
private var isVisited: Array<BooleanArray>? = null
private val dx = intArrayOf(-1, 0, 1, 0)
private val dy = intArrayOf(0, -1, 0, 1)


fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    N = sc.nextInt()
    M = sc.nextInt()
    map = Array(N + 2) { IntArray(M + 2) }
    isVisited = Array(N + 2) { BooleanArray(M + 2) }
    for (i in 0 until N) {
        val line = sc.next()
        for (j in 0 until line.length) {
            val num = line[j] - '0'
            map!![i + 1][j + 1] = num
        }
    }
    bfs(1, 1)
    println(map!![N][M])
}

fun bfs(x: Int, y: Int) {
    val queue = LinkedList<Point>()
    queue.offer(Point(x, y))
    isVisited!![x][y] = true
    while (!queue.isEmpty()) {
        val point = queue.poll()
        for (i in 0..3) {
            val x2 = point.x + dx[i]
            val y2 = point.y + dy[i]
            if (map!![x2][y2] == 1 && !isVisited!![x2][y2]) {
                isVisited!![x2][y2] = true
                queue.offer(Point(x2, y2))
                map!![x2][y2] = map!![point.x][point.y] + 1
            }
        }
    }
}

data class Point(var x: Int, var y: Int)



